src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -285,54 +285,164 @@ public static unsafe void MatMul(float[] mat, float[] src, float[] dst, int crow
+                            x2 = Sse.Multiply(x2, x1);
+                            result = Sse.Add(result, x2);
+                            ppos++;
+                        }
+
+                        Sse.Store(pDstCurrent, result);
+                        pDstCurrent += 4;
+                        pm0 += 4 * ccol;
+                    }
+                }
+                else
+                {
+                    if (misalignment != 0)
+                    {
+                        misalignment >>= 2;
+                        misalignment = 4 - misalignment;
+
+                        Vector128<float> mask = Sse.LoadVector128(((float*)(pLeadingAlignmentMask)) + (misalignment * 4));
 
-                    int* ppos = pposMin;
+                        float* pm1 = pm0 + ccol;
+                        float* pm2 = pm1 + ccol;
+                        float* pm3 = pm2 + ccol;
+                        Vector128<float> result = Sse.SetZeroVector128();
 
-                    while (ppos < pposEnd)
+                        int* ppos = pposMin;
+
+                        while (ppos < pposEnd)
+                        {
+                            int col = *ppos;
+                            Vector128<float> x1 = Sse.SetVector128(pm3[col], pm2[col], pm1[col], pm0[col]);
+
+                            x1 = Sse.And(mask, x1);
+                            Vector128<float> x2 = Sse.SetAllVector128(pSrcCurrent[col]);
+                            x2 = Sse.Multiply(x2, x1);
+                            result = Sse.Add(result, x2);
+                            ppos++;
+                        }
+
+                        Sse.Store(pDstCurrent, result);
+                        pDstCurrent += misalignment;
+                        pm0 += misalignment * ccol;
+                        length -= misalignment;
+                    }
+
+                    if (length > 3)
                     {
-                        int col = *ppos;
-                        Vector128<float> x1 = Sse.SetVector128(pm3[col], pm2[col], pm1[col], pm0[col]);
-                        Vector128<float> x2 = Sse.SetAllVector128(pSrcCurrent[col]);
-                        x2 = Sse.Multiply(x2, x1);
-                        result = Sse.Add(result, x2);
+                        remainder = length % 4;
+                        while (pDstCurrent < pDstEnd)
+                        {
+                            float* pm1 = pm0 + ccol;
+                            float* pm2 = pm1 + ccol;
+                            float* pm3 = pm2 + ccol;
+                            Vector128<float> result = Sse.SetZeroVector128();
+
+                            int* ppos = pposMin;
 
-                        ppos++;
+                            while (ppos < pposEnd)
+                            {
+                                int col = *ppos;
+                                Vector128<float> x1 = Sse.SetVector128(pm3[col], pm2[col], pm1[col], pm0[col]);
```

Comment:
note to self: I want to check the codegen of this and ensure that it is being emitted "optimally" (two loads and three unpack with two folded loads; rather than as four loads and three unpack).